### ISSUE TITLE
Flask-Security doesn't recover from error

### ISSUE THREAD

**Issue by Avamander** [mattupstate/flask-security#754]
For some reason my webpage one day decided to take itself down for some reason, giving me these stacktraces:

```
(psycopg2.OperationalError) SSL connection has been closed unexpectedly
 [SQL: 'SELECT "user".id AS user_id, "user".email AS user_email, "user".username AS user_username, "user".password AS user_password, "user".active AS user_active, "user".confirmed_at AS user_confirmed_at \nFROM "user" \nWHERE "user".id = %(id_1)s \n LIMIT %(param_1)s'] [parameters: {'param_1': 1, 'id_1': '0'}] (Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
psycopg2.OperationalError: SSL connection has been closed unexpectedly

Traceback (most recent call last):
  File "/home/avamander/.local/lib/python3.5/site-packages/raven/contrib/flask.py", line 238, in before_request
    self.client.user_context(self.get_user_info(request))
  File "/home/avamander/.local/lib/python3.5/site-packages/raven/contrib/flask.py", line 163, in get_user_info
    is_authenticated = current_user.is_authenticated
  File "/home/avamander/.local/lib/python3.5/site-packages/werkzeug/local.py", line 347, in __getattr__
    return getattr(self._get_current_object(), name)
  File "/home/avamander/.local/lib/python3.5/site-packages/werkzeug/local.py", line 306, in _get_current_object
    return self.__local()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/utils.py", line 26, in <lambda>
    current_user = LocalProxy(lambda: _get_user())
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/utils.py", line 335, in _get_user
    current_app.login_manager._load_user()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/login_manager.py", line 359, in _load_user
    return self.reload_user()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/login_manager.py", line 321, in reload_user
    user = self.user_callback(user_id)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_security/core.py", line 221, in _user_loader
    return _security.datastore.find_user(id=user_id)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_security/datastore.py", line 254, in find_user
    return self.user_model.query.filter_by(**kwargs).first()
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2778, in first
    ret = list(self[0:1])
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2570, in __getitem__
    return list(res)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2878, in __iter__
    return self._execute_and_instances(context)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2901, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
    raise value.with_traceback(tb)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) SSL connection has been closed unexpectedly
 [SQL: 'SELECT "user".id AS user_id, "user".email AS user_email, "user".username AS user_username, "user".password AS user_password, "user".active AS user_active, "user".confirmed_at AS user_confirmed_at \nFROM "user" \nWHERE "user".id = %(id_1)s \n LIMIT %(param_1)s'] [parameters: {'param_1': 1, 'id_1': '0'}] (Background on this error at: http://sqlalche.me/e/e3q8)
(sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back [SQL: 'SELECT "user".id AS user_id, "user".email AS user_email, "user".username AS user_username, "user".password AS user_password, "user".active AS user_active, "user".confirmed_at AS user_confirmed_at \nFROM "user" \nWHERE "user".id = %(id_1)s \n LIMIT %(param_1)s'] [parameters: [{}]]
Traceback (most recent call last):
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1125, in _execute_context
    conn = self._revalidate_connection()
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 427, in _revalidate_connection
    "Can't reconnect until invalid "
sqlalchemy.exc.InvalidRequestError: Can't reconnect until invalid transaction is rolled back

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/app.py", line 1610, in full_dispatch_request
    rv = self.preprocess_request()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask/app.py", line 1831, in preprocess_request
    rv = func()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_principal.py", line 477, in _on_before_request
    identity = loader()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_security/core.py", line 245, in _identity_loader
    if not isinstance(current_user._get_current_object(), AnonymousUserMixin):
  File "/home/avamander/.local/lib/python3.5/site-packages/werkzeug/local.py", line 306, in _get_current_object
    return self.__local()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/utils.py", line 26, in <lambda>
    current_user = LocalProxy(lambda: _get_user())
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/utils.py", line 335, in _get_user
    current_app.login_manager._load_user()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/login_manager.py", line 359, in _load_user
    return self.reload_user()
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_login/login_manager.py", line 321, in reload_user
    user = self.user_callback(user_id)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_security/core.py", line 221, in _user_loader
    return _security.datastore.find_user(id=user_id)
  File "/home/avamander/.local/lib/python3.5/site-packages/flask_security/datastore.py", line 254, in find_user
    return self.user_model.query.filter_by(**kwargs).first()
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2778, in first
    ret = list(self[0:1])
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2570, in __getitem__
    return list(res)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2878, in __iter__
    return self._execute_and_instances(context)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2901, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 948, in execute
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1132, in _execute_context
    None, None)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
    raise value.with_traceback(tb)
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1125, in _execute_context
    conn = self._revalidate_connection()
  File "/home/avamander/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 427, in _revalidate_connection
    "Can't reconnect until invalid "

Repeats indefinetly...
```
Any ideas how to avoid situations like these in the future?

**Comment by biomap** [comment]
This appears to be a psycopg/SQL/postgres error which I assume you have setup using SSL from the traceback you provided  >> psycopg2.OperationalError: SSL connection has been closed unexpectedly.   You need to call db.rollback() once you fix it and get your database connection reconnected.

**Comment by Avamander** [comment]
I encountered the error once again today. The thing is that the application is on the same server as my DB so I'm not sure why this connection dropping would even happen (idle timeout?), another thing is that my code isn't even executed when the error occurs, I can't detect and rollback it properly, Flask-Security should do so as it's its query that's failing.

**Comment by biomap** [comment]
Well to be honest the error is occurring with your database connection.  Everything is fine with FS as long as it has the DB access to perform the query. Its failing because it doesn't have access to the database to perform the select query or some other query is failing before the select query is issued.  I would dig into your logs and find out what the root of the problem is.  Select queries shouldn't require rollbacks as far as I'm aware even if they fail.

**Comment by biomap** [comment]
If you cant trace the problem you could try to wrap your login (where ever its failing) into a try statement but as I say the issue is with psycopg/sqlalchemy and your database connection.

**Comment by Avamander** [comment]
I wouldn't be surprised if it's just the SQL connection timeouting or some automated update (Ubuntu's) closing the connection, but even then, it shouldn't leave the entire app in an unusable state. 

Maybe you can read some solution out of these issues because I couldn't:
https://gitlab.com/mailman/mailman/issues/313
https://bitbucket.org/zzzeek/sqlalchemy/issues/3919

The worst case I'm just going to modify flask-security itself to try-catch that block and rollback and absolute worst is just restarting workers every once in a while.

**Comment by biomap** [comment]
Are you using this version of PG?  9.3.x?  If so maybe try to upgrade to the latest version of postgresql which is 10.x. https://www.postgresql.org/download/linux/ubuntu/

If its not in production and you are not worried about data loss you could just do it and uninstall the other or do a dump of the data and retain or you could run an additional pg server side by side on another port.

It appears to be related to your issue however as far as I can tell.  Also the newer versions of PG have some sweet new features that I use and you could find useful as well.

**Comment by biomap** [comment]
Hmm.... well reading more into that sqla thread it appears they have implemented something within sqlalchemy itself to resolve that issue.  It maybe a case of just upgrading to the latest version of sqla that would do it.  Also I saw in the same thread that Mike suggests disabling pooling entirely but that would be up to you.

If it were me I would upgrade my sqla version to the latest and see if that fixes your issue first then look at the connection pooling settings and test that if it didn't.  That would be the simplest approach.  I would still look at the newer versions of PG however as mentioned it has some really terrific new features from 9.4.x on.

Hope this helps!

**Comment by jirikuncar** [comment]
I am closing this issue as it doesn't look like an error in Flask-Security. Please link to the corresponding external issue to help other users. Thanks!

**Comment by Avamander** [comment]
I have updated literally everything in the setup but the issue still sometimes occurs. Flask-Security should auto-recover from **a query failing that it executes**. It's frankly quite annoying I have to start monkey-patching a library to recover from an error it creates.

**Comment by Avamander** [comment]
This is most likely caused by the DB connection silently timing out due to worker not using it for a while because there are a few timeout error stacktraces in my crash reporting.

**Comment by jwag956** [comment]
I concur that this isn't a FS issue per-se. All that should be handled at lower levels (and appears to be).
I had a similar issue with a bad query - and got the same traceback - but it only affected the one request - future requests seemed to work (which is expected since Flask-SqlAlchemy attaches and closes a session for each request). You imply that once this happens the app is dead?

Also - it is hard to infer from the stack trace - but I assume you are using a relatively 'stock' configuration - using FlaskSqlAlchemy for session mgmt and deriving your UserModel from SqlAlchemyUserDataStore (and not SqlAlchemySessionUserDataStore).

that being said - I think a simple overall solution to make sure this doesn't actually happen - is to utilize a new-ish (2017) feature of SqlAlchemy - outlined here:
https://docs.sqlalchemy.org/en/13/core/pooling.html#dealing-with-disconnects

Ideally this would be the default - but it isn't, furthermore FlaskSqlAlchemy has an issue described here:
https://github.com/pallets/flask-sqlalchemy/pull/67
that makes it hard to add custom options to create_engine. Fortunately, in that discussion - there is a very easy workaround:

`class MySQLAlchemy(SQLAlchemy):
    def apply_driver_hacks(app, info, options):
        options.update(pool_pre_ping=True)
        super(MySQLAlchemy, self).apply_driver_hacks(app, info, options)

db = MySQLAlchemy(app)`

**Comment by Avamander** [comment]
>  You imply that once this happens the app is dead?

The same "Can't reconnect until invalid transaction is rolled back" appears in many different components that relate to Flask-Security, meaning it's not exactly dead, but not very usable either.

As long as pool pre-ping isn't the default, Flask-Security should make sure the pool is actually usable (or fix when erroring) before taking itself into a broken state, that's my opinion at least.

**Comment by flowswitch** [comment]
I had a similar issue ("Lost connection to DB", "Can't reconnect until invalid transaction is rolled back") with a code based on this example: https://pythonhosted.org/Flask-Security/quickstart.html#id2

My hoster's (pythonanywhere) MySQL server is configured for 5 min idle connection timeout, so they advise passing an additional 'pool_recycle=280' parameter to SQLAlchemy's create_engine, people at their discussion confirmed that it solves the issue, but it wasn't working for me at all. 

Finally I've ran into some discussion mentioning that importing SQLAlchemy directly has some ill effects with Flask,  stating that flask_sqlalchemy wrapper module should be used instead. So I've reworked my code to something closer to https://pythonhosted.org/Flask-Security/quickstart.html#basic-sqlalchemy-application, set the pool_recycle param in app.config['SQLALCHEMY_ENGINE_OPTIONS'] and the problem have gone.
